pkg/apis/workflow/v1alpha1/data_types.go

```diff
@@ -0,0 +1,37 @@
+package v1alpha1
+
+// Data is a data template
+type Data struct {
+	// Source sources external data into a data template
+	Source DataSource `json:"source" protobuf:"bytes,1,opt,name=source"`
+
+	// Transformation applies a set of transformations
+	Transformation Transformation `json:"transformation" protobuf:"bytes,2,rep,name=transformation"`
+}
+
+func (d *Data) GetArtifactIfAny() *Artifact {
```

Comment:
the abstraction feels wrong here, artifacts do not make sense for other sources - and they ought to if you have this func

Reply:
Updated: this is a remnant of a previous design